services/QuillLMS/app/models/concept_result.rb

```diff
@@ -0,0 +1,145 @@
+#  answer                              :jsonb
+#  attempt_number                      :integer
+#  correct                             :boolean          not null
+#  extra_metadata                      :jsonb
+#  question_number                     :integer
+#  question_score                      :float
+#  created_at                          :datetime         not null
+#  activity_session_id                 :integer          not null
+#  concept_id                          :integer
+#  concept_result_directions_id        :integer
+#  concept_result_instructions_id      :integer
+#  concept_result_previous_feedback_id :integer
+#  concept_result_prompt_id            :integer
+#  concept_result_question_type_id     :integer
+#  old_concept_result_id               :integer
+#
+# Indexes
+#
+#  index_concept_results_on_old_concept_result_id  (old_concept_result_id) UNIQUE
+#
+class ConceptResult < ApplicationRecord
+  belongs_to :activity_session
+  belongs_to :concept
+  belongs_to :old_concept_result
+  belongs_to :concept_result_directions
+  belongs_to :concept_result_instructions
+  belongs_to :concept_result_previous_feedback
+  belongs_to :concept_result_prompt
+  belongs_to :concept_result_question_type
+
+  validates_inclusion_of :correct, in: [true, false]
+
+  # This is a list of keys from the old ConceptResults records
+  # that this model knows how to normalize.  Any keys that are
+  # fed into create_from_json that aren't from this list,
+  # will be stashed in a related extra_metadata
+  # record
+  KNOWN_METADATA_KEYS = [
+    :answer,
+    :attemptNumber,
+    :correct,
+    :directions,
+    :lastFeedback,
+    :prompt,
+    :questionNumber,
+    :questionScore
+  ]
+
+  def self.create_from_json(data_hash)
+    response = init_from_json(data_hash)
+    response.save!
+    response
+  end
+
+  def self.init_from_json(data_hash)
+    data_hash = data_hash.deep_symbolize_keys
+
+    metadata = data_hash[:metadata]
+
+    response = new(
+      activity_session_id: data_hash[:activity_session_id],
+      answer: metadata[:answer],
+      concept_id: data_hash[:concept_id],
+      attempt_number: metadata[:attemptNumber],
+      correct: ActiveModel::Type::Boolean.new.cast(metadata[:correct]),
+      question_number: metadata[:questionNumber],
+      question_score: metadata[:questionScore]
+    )
+
+    response.assign_normalized_text(data_hash)
+    response.assign_extra_metadata(metadata)
+
+    response
+  end
+
+  def self.bulk_create_from_json(data_hash_array)
+    data_hash_array.map { |data_hash| create_from_json(data_hash) }
+  end
+
+  def self.find_or_create_from_old_concept_result(old_concept_result)
+    return old_concept_result.concept_result if old_concept_result.concept_result
+
+    create_from_json(old_concept_result.as_json)
+  end
+
+  def self.parse_extra_metadata(metadata)
+    metadata.deep_symbolize_keys.except(*KNOWN_METADATA_KEYS).reject{ |_,v| v.nil? || (v.is_a?(Enumerable) && v.empty?) }
```

Comment:
Let's put code like this on multiple lines for readability
```suggestion
    metadata
      .deep_symbolize_keys
      .except(*KNOWN_METADATA_KEYS)
      .reject{ |_,v| v.nil? || (v.is_a?(Enumerable) && v.empty?) }
```